runtime/norm/src/.component/.expand/version.js

```diff
@@ -1,19 +1,46 @@
 'use strict'
 
-function version (manifest) {
-  if (manifest.version === undefined) {
-    const bridge = require(manifest.bridge)
+const { join } = require('node:path')
+const { createHash } = require('node:crypto')
+const fs = require('node:fs/promises')
+const { createReadStream } = require('node:fs')
+const { once } = require('node:events')
 
-    if ('version' in bridge)
-      manifest.version = bridge.version(manifest)
-  }
+async function version (manifest) {
+  manifest.version ??= await hash(manifest.path)
+}
+
+async function hash (path) {
+  const hash = await hashd(path)
+  const hex = hash.digest('hex')
+
+  return hex.slice(0, 8)
+}
 
-  if (manifest.version === undefined) {
-    console.warn(`Component '${manifest.namespace ? manifest.namespace + '.' : ''}${manifest.name}' has no version`)
+/**
+ * @param {string} path
+ * @param {import('node:crypto').Hash} hash
+ */
+async function hashd (path, hash = createHash('sha256')) {
+  const stat = await fs.stat(path)
 
-    manifest.version = Math.random().toString(36).slice(2)
+  if (stat.isFile()) {
+    const stream = createReadStream(path)
+
+    stream.pipe(hash, { end: false })
+
+    return await once(stream, 'end')
   }
 
+  if (stat.isDirectory()) {
+    const entries = await fs.readdir(path)
+
+    for (const entry of entries) {
+      await hashd(join(path, entry), hash)
+    }
+
```

Comment:
This implementation will completely ignore symlinks (`isSymbolicLink()` case) and that is how some packages managers like `pnpm` or `yarn` are installing dependencies nowadays. Is that by design?

Reply:
This reveals deeper problem.

While components usually developer inside a monorepo, thus their dependencies are pinned in the root level `package-lock`, when they are [installed inside docker image](https://github.com/toa-io/toa/blob/f28d629a9477646e267a8af8479cc1bb10d62c80/operations/src/deployment/images/composition.Dockerfile#L16), this `package-lock` doesn't exists.

At this point I think we may assume that if there are no changes in `package.json`, then previously built image may be used for deployment.

Unless you have some other suggestions.

Reply:
I do agree that it probably OK for our case. People with complex deps can either pin exact dependency version or manage component versioning with lerna.